### PR TITLE
ONFRONT-25 bypass app name question

### PR DESCRIPTION
**Pull request by matt-stjean**
This PR introduces several enhancements and small refactorings to the Ontario.ca Frontend CLI tool, specifically addressing the ability to specify the application name via a CLI parameter. 

## Included Changes

1. Users can now specify their desired application name using the --projectName option when initializing a new project. This bypasses the corresponding question in the interactive setup, speeding up the initialization process.

Example usage:

```sh
npx @ongov/ontario-frontend ontario-create-app --projectName my-frontend-project
```

2. The CLI tool validates the provided application name to ensure it meets Ontario.ca naming conventions (lowercase, hyphens, underscores allowed).

Validation is performed using the validFileName function from fileNameValidator.js.

3. Consolidated how options and answers are handled in the handleCreateAppCommand function for improved readability and maintainability.

4. Rearranged the question logic to ensure unanswered questions are properly prompted.

5. Introduced a debug mode to ontario-create-app assist in troubleshooting and development, using the `--debug` option. 

Example usage: 

```sh
npx @ongov/ontario-frontend ontario-create-app --projectName my-frontend-project --debug
```

7. Added an informational log message to confirm when the project name is bypassed using the CLI parameter.

## Acceptance Criteria

- The user can specify the application name using a CLI parameter.
- The interactive setup is bypassed when the parameter is provided.
- The application name is validated to follow naming conventions.
- If no name is provided, the interactive setup prompts for the application name.
- Confirmation messages and error handling have been implemented to ensure a smooth user experience.

### CODE REVIEW

File: packages/ontario-frontend-cli/core/utils/logger.js [hunk 1]
```diff
@@ -1,10 +1,46 @@
...
+const formatMessages = (...messages) => {
+  return messages
+    .map((message) => {
+      if (typeof message === 'object') {
+        return JSON.stringify(message, null, 2);
+      }
+      return message;
+    })
+    .join(' ');
+};
```
Comment (kyle-binger-3) [inline packages/ontario-frontend-cli/core/utils/logger.js:29]: Nice!

File: package.json [hunk 2]
```diff
@@ -20,5 +20,5 @@
     "node": ">=18.17.0",
     "pnpm": ">=8.15.1"
   },
-  "packageManager": "^pnpm@8.15.1"
```
Comment (kyle-binger-3) [inline package.json:23]: I think we keep going back and forth on this caret, I was able to run `pnpm` commands with the caret on my end, are you able to?

File: packages/ontario-frontend-cli/commands/ontario-create-app/handleCreateApp.js [hunk 3]
```diff
@@ -25,49 +25,66 @@ const { createOntarioAppQuestions } = require('../../core/questions');
...
+      logger.debug('Prettier configuration copied successfully');
     } catch (err) {
-      throw err
+      throw new Error(`Failed to copy Prettier configuration: ${err.message}`);
     }
   }
 
   // Install npm dependencies, including the core Frontend dependency
   await installAllPackages(newProjectPath);
+  logger.debug('npm dependencies installed successfully');
```
Comment (kyle-binger-3) [inline packages/ontario-frontend-cli/commands/ontario-create-app/handleCreateApp.js]: Very minor but we seem to capitalize NPM in the success message elsewhere, the debug message should probably match the capitalized sentence case.
Screenshot 2024-06-06 at 2 14 05 PM

**Comment by matt-stjean** [comment]
Re: new commits

- Cleaned up the ontario-create-app command code in ontario-create-app.js
- Investigated potential defect with validation which found that $ and ! are reserved bash characters and not something we can effectively handle. However, investigation revealed that the validation regex could be improved
- Copied and Pasted the CLI description from confluence into the package.json

File: packages/ontario-frontend-cli/commands/ontario-create-app/handleCreateApp.js [hunk 4]
```diff
@@ -25,49 +25,66 @@ const { createOntarioAppQuestions } = require('../../core/questions');
...
 
   // Copy ESLint config if opted-in
   if (conf.addESLint) {
     try {
       await handlePackageCopy(newProjectPath, 'eslint');
+
+      logger.debug('ESLint configuration copied successfully');
     } catch (err) {
-      throw err
+      throw new Error(`Failed to copy ESLint configuration: ${err.message}`);
```
Comment (kyle-binger-3) [inline packages/ontario-frontend-cli/commands/ontario-create-app/handleCreateApp.js:70]: Definitely a friendly error message than before, same with prettier below, nice catch!

File: packages/ontario-frontend-cli/core/validation/validationUtils.js [hunk 5]
```diff
@@ -5,7 +5,7 @@ const isPresent = (value) => value && value.trim() !== '';
 const isLowercase = (value) => value === value.toLowerCase();
 
 // Checks if the value is URL-safe (alphanumeric, hyphens, underscores)
-const isUrlSafe = (value) => /^[\w-]+$/.test(value);
+const isUrlSafe = (value) => /^[a-z0-9_-]+$/.test(value);
```
Comment (kyle-binger-3) [inline packages/ontario-frontend-cli/core/validation/validationUtils.js:8]: Nice improvement! It's a bit more specific.

File: packages/ontario-frontend-cli/package.json [hunk 6]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "@ongov/ontario-frontend-cli",
   "version": "1.1.0",
-  "description": "",
+  "description": "The Ontario.ca Frontend Command Line Interface (CLI) is a series of tools for initializing, managing and developing Ontario.ca Frontend projects. It provides a simple interface to scaffold new projects, configure them with essential dependencies, and prepare them for development using Ontario.ca Frontend.",
```
Comment (kyle-binger-3) [inline packages/ontario-frontend-cli/package.json:4]: Nice catch and good description 👍

**Review by kyle-binger-3 — COMMENTED** [review]
Awesome job Matt! Great work, really like the addition of the debug information, especially the `ontarioFrontendDependency` field.

File: package.json (hunk 2 above)
Comment (kyle-binger-3) [reply package.json:23]: Looks like the caret is unsupported syntax, we can remove it
https://stackoverflow.com/questions/71747609/how-to-specify-packagemanager-in-package-json

**Review by kyle-binger-3 — DISMISSED** [review]

File: packages/ontario-frontend-cli/commands/ontario-create-app/handleCreateApp.js (hunk 3 above)
Comment (mattstjean) [reply packages/ontario-frontend-cli/commands/ontario-create-app/handleCreateApp.js]: Removing the message from here as it provides no value.

**Review by kyle-binger-3 — APPROVED** [review]

**Pull request merged**